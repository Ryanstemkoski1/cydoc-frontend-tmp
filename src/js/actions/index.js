import {ADD_NOTE, ADD_TEMPLATE, DATA_LOADED, LOGIN_REQUEST, LOGOUT, SAVE_NOTE} from "../constants/action-types";
import axios from 'axios'
import api from "../constants/api";

export function addNote(payload) {
    return { type: ADD_NOTE, payload };
}

export function addTemplate(payload) {
    return { type: ADD_TEMPLATE, payload: payload}
}

export function getAllRecords() {
    return function(dispatch) {
        return axios.get(api.records.dev)
            .then(response => response.data)
            .then(data => {
                dispatch({ type: DATA_LOADED, payload: data });
            });
    };
}

export function loginRequest(payload){
    return function(dispatch) {
        return axios.post(api.login.dev, payload)
            .then(res => {
                const user = res;
                console.log(user)
                localStorage.setItem('user', JSON.stringify(user));
                return user;
            })
            .then((user)=> {
                dispatch({type: LOGIN_REQUEST, payload: user.data})
                return user;
            })
            .catch(err => {
                return err.response;
            })
    }
}

export function logout(){
        localStorage.removeItem('user');
        return { type: LOGOUT };
}

export  function saveNote(payload){
    return function (dispatch) {
        const headers = {
            'Content-Type': 'application/json',

        };
        return axios.post(api.newRecord.dev,
            { headers: {
                'Content-Type': 'application/json'}},
            payload)
            .then(res => res.data)
            .then(data => {
                dispatch({type: SAVE_NOTE, payload: data})
            })
    }
}