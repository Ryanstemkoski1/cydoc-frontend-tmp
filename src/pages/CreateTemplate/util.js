import { questionTypes } from '../../constants/questionTypes';

// Filler text for when the root is NaN
export const NAN_QUESTION_TEXT =
    '(This question has no text, only follow-up questions)';

/**
 * Returns starting answerInfo object associated with the given type
 * @param {String} type: the question's response type
 */
export const getAnswerInfo = (type) => {
    let answerInfo;
    switch (type) {
        case questionTypes.YES_NO:
        case questionTypes.NO_YES:
            answerInfo = {
                yesResponse: '',
                noResponse: '',
            };
            break;
        case questionTypes.SHORT_TEXT:
        case questionTypes.NUMBER:
        case questionTypes.TIME:
        case questionTypes.BODY_LOCATION:
        case questionTypes.LIST_TEXT:
            answerInfo = {
                startResponse: '',
                endResponse: '',
            };
            break;
        case questionTypes.CLICK_BOXES:
            answerInfo = {
                startResponse: '',
                endResponse: '',
                options: ['', '', '', ''],
            };
            break;
        default:
            // Advanced type questions
            answerInfo = {
                options: [],
            };
    }
    return answerInfo;
};

/**
 * Returns a randomly generated node qid based on the disease code
 * @param {String} diseaseCode: 3-character abbreviation of the disease
 * @param {*} numQuestions: number of questions in the context
 */
export const createNodeId = (diseaseCode, numQuestions) => {
    const randomId = Math.floor(Math.random() * 9000000000) + 1000000000;
    const numZeros = 4 - numQuestions.toString().length;
    return (
        diseaseCode +
        '-' +
        randomId.toString() +
        '-' +
        '0'.repeat(numZeros) +
        numQuestions.toString()
    );
};

/**
 * Sorts the list of edges in place based on the edge's `toQuestionOrder`
 * @param {Array[Object]} edgeList
 * @param {Array[Object]} edges
 */
export const sortEdges = (edgeList, edges) => {
    edgeList.sort((a, b) => {
        const nodeA = parseFloat(edges[a].toQuestionOrder);
        const nodeB = parseFloat(edges[b].toQuestionOrder);

        if (nodeA < nodeB) {
            return -1;
        } else if (nodeA > nodeB) {
            return 1;
        } else {
            return 0;
        }
    });
};

/**
 * Given the node's responseType, text, and answerInfo, parse the text for the options,
 * if appropriate, and mutate the node's answerInfo directly.
 * Returns the parsed text.
 *
 * @param {String} responseType
 * @param {String} text
 * @param {Object} answerInfo
 * @param {String} category
 */
export const parseQuestionText = (responseType, text, answerInfo, category) => {
    if (text === 'nan') {
        text = `Root for ${category
            .replace('_', ' ')
            .toLowerCase()} ${NAN_QUESTION_TEXT}`;
    }

    if (
        responseType === questionTypes.CLICK_BOXES ||
        responseType.endsWith('POP') ||
        responseType === 'nan'
    ) {
        let click = text.search('CLICK');
        let selectStart = text.search('\\[');
        let selectEnd = text.search('\\]');
        let choices;
        if (click > -1) {
            // options are indicated by CLICK[...]
            choices = text.slice(click + 6, selectEnd);
            text = text.slice(0, click);
        } else {
            // options are indicated by [...]
            if (selectStart > 0) {
                choices = text.slice(selectStart + 1, selectEnd);
                text = text.slice(0, selectStart);
            }
        }
        choices = choices.split(',').map((response) => response.trim());
        answerInfo.options = choices;
    }
    return text;
};

/**
 * Replace all occurences of SYMPTOM or DISEASE with the given category
 *
 * @param {String} text
 * @param {String} category
 */
export const parsePlaceholder = (text, category) => {
    let placeholder = text.search(/SYMPTOM|DISEASE/);
    if (placeholder > -1) {
        text =
            text.substring(0, placeholder) +
            category.replace('_', ' ').toLowerCase() +
            text.substring(placeholder + 7);
    }
    return text;
};

/**
 * Adds blank{Yes, No, Template} to the appropriate fields in answerInfo
 */
export const setAnswerInfo = (answerInfo, node) => {
    if ('yesResponse' in answerInfo) {
        answerInfo.yesResponse = node.blankYes || '';
        answerInfo.noResponse = node.blankNo || '';
    } else if ('startResponse' in answerInfo) {
        const capturingRegex = /(?<startResponse>.*)\bANSWER\b((?<negStartResponse>.*)(\bNOTANSWER\b(?<negEndResponse>.*))|(?<endResponse>.*))/;
        const found = node.blankTemplate.match(capturingRegex);
        if (found) {
            answerInfo.startResponse = found.groups.startResponse || '';
            if (found.groups.endResponse) {
                answerInfo.endResponse = found.groups.endResponse || '';
            } else {
                answerInfo.endResponse = found.groups.negStartResponse || '';
                answerInfo.negEndResponse = found.groups.negEndResponse || '';
            }
        }
    }
    return answerInfo;
};

/**
 * Adds all direct children question of the given parent directly
 * to the graph object itself.
 *
 * Returns the updated numQuestions and nextEdgeID count.
 */
export const addChildrenNodes = (
    parentId,
    children,
    category,
    diseaseCode,
    graphData,
    contextData
) => {
    let {
        numQuestions,
        nextEdgeID,
        contextEdges,
        contextGraph,
        contextNodes,
    } = contextData;
    const { graph, edges, nodes } = graphData;

    // Create edges and nodes for every new question
    for (let edge of children) {
        const childId = createNodeId(diseaseCode, numQuestions);
        const nodeId = edges[edge].to;
        let responseType = nodes[nodeId].responseType;
        let text = nodes[nodeId].text;
        let answerInfo = getAnswerInfo(responseType);
        answerInfo = setAnswerInfo(answerInfo, nodes[nodeId]);

        // Preprocess the text to prepopulate the answerinfo if necessary
        // or replace instances of SYMPTOM and DISEASE
        text = parsePlaceholder(text, category);
        text = parseQuestionText(
            responseType,
            text,
            answerInfo,
            nodes[nodeId].category
        );
        contextNodes[childId] = {
            text,
            answerInfo,
            responseType,
            id: childId,
            order: numQuestions,
            hasChildren: graph[nodeId].length > 0,
            originalId: nodeId,
            hasChanged: false,
            parent: parentId,
        };

        contextEdges[nextEdgeID] = {
            from: parentId,
            to: childId,
        };

        contextGraph[childId] = [];
        contextGraph[parentId].push(nextEdgeID);
        nextEdgeID++;
        numQuestions++;
    }
    return { nextEdgeID, numQuestions };
};

/**
 * Walk up the knowledge graph and update all imported nodes
 * to hasChanged=true to indicate not to use the original ID anymore
 *
 * @param {Object} nodes
 * @param {String} qid
 */
export const updateParent = (nodes, qid) => {
    if (qid && !nodes[qid].hasChanged) {
        nodes[qid].hasChanged = true;
        updateParent(nodes, nodes[qid].parent);
    }
};

// Returns blank template for HPITemplateContext
export const getNewTemplate = () => {
    const rootId = createNodeId('000', 0);
    return {
        // `null` graphID indicates this is a newly created graph
        graphID: null,
        title: '',
        disease: '',
        bodySystem: '',
        numQuestions: 1,
        nextEdgeID: 0,
        hasChanged: true,
        root: rootId,
        graph: {
            [rootId]: [],
        },
        nodes: {
            [rootId]: {
                id: rootId,
                responseType: 'YES-NO',
                text: 'nan',
            },
        },
        edges: {},
        cydocGraphs: { graph: {}, edges: {}, nodes: {} },
    };
};
