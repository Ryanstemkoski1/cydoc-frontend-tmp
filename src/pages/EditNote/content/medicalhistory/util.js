/**
 * From https://stackoverflow.com/questions/38345273/javascript-algorithm-converting-to-roman-numbers
 */
const toRomanNumeral = (num) => {
    var arrConv = {
        0: { 1: 'I', 2: 'II', 3: 'III', 4: 'IV', 5: 'V', 6: 'VI', 7: 'VII', 8: 'VIII', 9: 'IX' },
        1: { 1: 'X', 2: 'XX', 3: 'XXX', 4: 'XL', 5: 'L', 6: 'LX', 7: 'LXX', 8: 'LXXX', 9: 'XC' },
        2: { 1: 'C', 2: 'CC', 3: 'CCC', 4: 'CD', 5: 'D', 6: 'DC', 7: 'DCC', 8: 'DCCC', 9: 'CM' },
        3: { 1: 'M', 2: 'MM', 3: 'MMM', 4: 'MMMM', 5: 'MMMMM', 6: 'MMMMMM', 7: 'MMMMMMM', 8: 'MMMMMMMM', 9: 'MMMMMMMMM' }
    },
    arr = num.toString().split("").reverse(),
    romansArray = arr.map(function (a, i) {
        return arrConv[i][a] || '';
    });

    return romansArray.reverse().join("");
}

/**
 * Replaces all numbers to roman numerals and all chars to lower case
 */
export const adjustValue = (val, mapping) => {
    let value = val.toLowerCase();
    if (value in mapping) {
        return mapping[value];
    }
    return val
        .replace(/\d+/gi, num => toRomanNumeral(num))
        .toLowerCase();
}
