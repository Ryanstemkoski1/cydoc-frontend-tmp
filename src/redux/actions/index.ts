import { ROSActionTypes } from './reviewOfSystemsActions';
import { MedicalHistoryActionTypes } from './medicalHistoryActions';
import { SurgicalHistoryActionTypes } from './surgicalHistoryActions';
import { MedicationsActionTypes } from './medicationsActions';
import { AllergiesActionTypes } from './allergiesActions';
import { SocialHistoryActionTypes } from './socialHistoryActions';
import { FamilyHistoryActionTypes } from './familyHistoryActions';
import { PhysicalExamActionTypes } from './physicalExamActions';
import { PlanActionTypes } from './planActions';
import { HpiActionTypes } from './hpiActions';
import { WidgetActionTypes } from './widgetActions';
import { chiefComplaintsActionTypes } from './chiefComplaintsActions';

export type AllActionTypes =
    | ROSActionTypes
    | PhysicalExamActionTypes
    | WidgetActionTypes
    | MedicalHistoryActionTypes
    | SurgicalHistoryActionTypes
    | MedicationsActionTypes
    | AllergiesActionTypes
    | SocialHistoryActionTypes
    | FamilyHistoryActionTypes
    | PlanActionTypes
    | HpiActionTypes
    | chiefComplaintsActionTypes;
