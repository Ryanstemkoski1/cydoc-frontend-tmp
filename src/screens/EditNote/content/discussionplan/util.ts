/**
 * @fileoverview Miscellaneous helper functions and types/interfaces for the
 * discussion and plan page.
 */

import { PlanCondition } from '@redux/reducers/planReducer';
import { PlanActionTypes } from '@redux/actions/planActions';
import { WhenResponse, YesNoUncertainResponse } from '@constants/enums';

export interface DropdownOptions {
    value: string;
    label: string;
}

export type EventHandler = (_e: any, data: any) => void;

export type HandleOnAddItem = (_e: any, data: { [key: string]: any }) => void;

export type PlanAction = (
    conditionIndex: string,
    ...args: any[]
) => PlanActionTypes;

export type ConditionCategoryKey = keyof Omit<PlanCondition, 'name'>;

// Initial values with deterministic ids to allow for snapshot testing
export const conditionId = 'condition';
export const categoryId = 'category';
export const initialPlan = {
    conditions: {
        [conditionId]: {
            name: '',
            differentialDiagnoses: {
                [categoryId]: {
                    comments: '',
                    diagnosis: '',
                },
            },
            prescriptions: {
                [categoryId]: {
                    comments: '',
                    dose: '',
                    type: '',
                    signature: '',
                },
            },
            proceduresAndServices: {
                [categoryId]: {
                    comments: '',
                    procedure: '',
                    when: WhenResponse.None,
                },
            },
            referrals: {
                [categoryId]: {
                    comments: '',
                    department: '',
                    when: WhenResponse.None,
                },
            },
        },
    },
    survey: {
        admitToHospital: YesNoUncertainResponse.None,
        emergency: YesNoUncertainResponse.None,
        sicknessLevel: 0,
    },
};
