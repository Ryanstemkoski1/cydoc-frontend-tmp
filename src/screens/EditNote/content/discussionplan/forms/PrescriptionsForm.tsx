import { OptionMapping } from '../../../../../_processOptions';
import Dropdown from '@components/tools/OptimizedDropdown';
import React from 'react';
import { connect } from 'react-redux';
import {
    addPrescription,
    updatePrescriptionComments,
    updatePrescriptionDose,
    updatePrescriptionSignature,
    updatePrescriptionType,
} from '@redux/actions/planActions';
import { CurrentNoteState } from '@redux/reducers';
import {
    PlanPrescriptionFlat,
    selectPlanCondition,
} from '@redux/selectors/planSelectors';
import { Grid, TextArea } from 'semantic-ui-react';
import { PlanAction } from '../util';
import {
    BaseCategoryForm,
    CategoryFormComponent,
    CategoryFormOwnProps,
    CategoryFormProps,
} from './BaseCategoryForm';
import './DiscussionPlanForms.css';
import useUpdateDimensions from './useUpdateDimensions';
import './planSections.css';

interface PrescriptionsDispatchProps {
    addPrescription: PlanAction;
    updatePrescriptionComments: PlanAction;
    updatePrescriptionDose: PlanAction;
    updatePrescriptionType: PlanAction;
    updatePrescriptionSignature: PlanAction;
}

type ComponentFunction = CategoryFormComponent<PlanPrescriptionFlat>;

const PrescriptionsForm = (
    props: CategoryFormProps<PlanPrescriptionFlat> & PrescriptionsDispatchProps
) => {
    const { categoryData, formatAction, ...actions } = props;
    useUpdateDimensions();

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column> Rx </Grid.Column>
            <Grid.Column> Signature (Sig) </Grid.Column>
            <Grid.Column> Comments </Grid.Column>
        </Grid.Row>
    );

    const mainInput: ComponentFunction = (row, options, onAddItem) => (
        <div className='container' id='prescriptions-main-input-container'>
            <Dropdown
                fluid
                search
                selection
                allowAdditions
                clearable
                optiontype='main'
                transparent={false}
                uuid={row.id}
                value={row.type}
                options={(options?.main as OptionMapping) || {}}
                onChange={formatAction(actions.updatePrescriptionType)}
                onAddItem={onAddItem}
                aria-label='Prescription-Dropdown'
                placeholder='medication name'
                className='main-input recipe'
            />
            <div className='container' id='prescriptions-container3' />
            <label>Prescription Dose</label>
            <div className='container' id='prescriptions-container2' />
            <div className='ui form'>
                <TextArea
                    type='text'
                    uuid={row.id}
                    value={row.dose}
                    onChange={formatAction(actions.updatePrescriptionDose)}
                    aria-label='Prescription-Amount'
                    placeholder='e.g. 81 mg tablet'
                    className={`recipe-amount lg`}
                />
            </div>
        </div>
    );

    const gridColumn: ComponentFunction = (row, options, onAddItem) => (
        <React.Fragment key={row.id}>
            <Grid.Column>{mainInput(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionSignature
                        )}
                        value={row.signature}
                        aria-label='Prescription-Signature'
                        placeholder='e.g. 1 tablet every 8 hours'
                    />
                </div>
            </Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionComments
                        )}
                        value={row.comments}
                        aria-label='Prescription-Comment'
                        placeholder='e.g. take with food'
                    />
                </div>
            </Grid.Column>
        </React.Fragment>
    );

    return (
        <BaseCategoryForm<PlanPrescriptionFlat>
            category='prescriptions'
            categoryData={categoryData}
            numColumns={3}
            addRowLabel='prescription'
            // addRow={formatAction(actions.addPrescription)}
            addRow={() => actions.addPrescription(actions.conditionId)}
            components={{
                gridColumn,
                gridHeaders,
            }}
        />
    );
};

export default connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .prescriptions,
    }),
    {
        addPrescription,
        updatePrescriptionComments,
        updatePrescriptionDose,
        updatePrescriptionType,
        updatePrescriptionSignature,
    }
)(PrescriptionsForm);
