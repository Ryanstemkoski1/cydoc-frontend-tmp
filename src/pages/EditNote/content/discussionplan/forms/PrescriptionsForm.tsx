import React from 'react';
import { Grid, TextArea } from 'semantic-ui-react';
import Dropdown from 'components/tools/OptimizedDropdown';
import { PlanAction } from '../util';
import { connect } from 'react-redux';
import { CurrentNoteState } from 'redux/reducers';
import {
    PlanPrescriptionFlat,
    selectPlanCondition,
} from 'redux/selectors/planSelectors';
import {
    addPrescription,
    updatePrescriptionComments,
    updatePrescriptionDose,
    updatePrescriptionType,
    updatePrescriptionSignature,
} from 'redux/actions/planActions';
import {
    CategoryFormProps,
    CategoryFormComponent,
    CategoryFormOwnProps,
    BaseCategoryForm,
} from './BaseCategoryForm';
import UpdateDimensions from './UpdateDimensions';
import './planSections.css';

interface PrescriptionsDispatchProps {
    addPrescription: PlanAction;
    updatePrescriptionComments: PlanAction;
    updatePrescriptionDose: PlanAction;
    updatePrescriptionType: PlanAction;
    updatePrescriptionSignature: PlanAction;
}

type ComponentFunction = CategoryFormComponent<PlanPrescriptionFlat>;

const PrescriptionsForm = (
    props: CategoryFormProps<PlanPrescriptionFlat> & PrescriptionsDispatchProps
) => {
    const { mobile, categoryData, formatAction, ...actions } = props;
    const { width } = UpdateDimensions();

    const gridHeaders = () => (
        <Grid.Row>
            <Grid.Column> Rx </Grid.Column>
            <Grid.Column> Signature (Sig) </Grid.Column>
            <Grid.Column> Comments </Grid.Column>
        </Grid.Row>
    );

    const mainInput: ComponentFunction = (row, options, onAddItem) => (
        <div className='container' id='prescriptions-main-input-container'>
            <Dropdown
                fluid
                search
                selection
                allowAdditions
                clearable
                optiontype='main'
                transparent={mobile}
                uuid={row.id}
                value={row.type}
                options={options?.main || {}}
                onChange={formatAction(actions.updatePrescriptionType)}
                onAddItem={onAddItem}
                aria-label='Prescription-Dropdown'
                placeholder='medication name'
                className='main-input recipe'
            />
            <div className='container' id='prescriptions-container3' />
            <label>Prescription Dose</label>
            <div className='container' id='prescriptions-container2' />
            <div className='ui form'>
                <TextArea
                    fluid
                    type='text'
                    transparent={mobile}
                    uuid={row.id}
                    value={row.dose}
                    onChange={formatAction(actions.updatePrescriptionDose)}
                    aria-label='Prescription-Amount'
                    placeholder='e.g. 81 mg tablet'
                    className={`recipe-amount ${!mobile && 'lg'}`}
                />
            </div>
        </div>
    );

    const gridColumn: ComponentFunction = (row, options, onAddItem) => (
        <React.Fragment key={row.id}>
            <Grid.Column>{mainInput(row, options, onAddItem)}</Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionSignature
                        )}
                        value={row.signature}
                        aria-label='Prescription-Signature'
                        placeholder='e.g. 1 tablet every 8 hours'
                    />
                </div>
            </Grid.Column>
            <Grid.Column>
                <div className='ui form'>
                    <TextArea
                        uuid={row.id}
                        onChange={formatAction(
                            actions.updatePrescriptionComments
                        )}
                        value={row.comments}
                        aria-label='Prescription-Comment'
                        placeholder='e.g. take with food'
                    />
                </div>
            </Grid.Column>
        </React.Fragment>
    );

    const mobileTitle: ComponentFunction = (row, options, onAddItem) => (
        <div className='recipe'>
            <label> Rx </label>
            {mainInput(row, options, onAddItem)}
        </div>
    );

    const mobileContent: ComponentFunction = (row) => (
        <>
            <label className='prescriptions-label'> Signature (Sig) </label>
            <div className='container' id='prescriptions-container1' />
            <div className='ui form' id='prescriptions-ui-form'>
                <TextArea
                    fluid
                    transparent
                    type='text'
                    uuid={row.id}
                    value={row.signature}
                    onChange={formatAction(actions.updatePrescriptionSignature)}
                    placeholder='e.g. 1 tablet every 8 hours'
                    aria-label='Prescription-Signature'
                    className='expanded-input'
                />
            </div>
            <div className='container' id='prescriptions-container2' />
            <label className='prescriptions-label'> Comments </label>
            <div className='container' id='prescriptions-container1' />
            <div className='ui form' id='prescriptions-ui-form'>
                <TextArea
                    fluid
                    transparent
                    type='text'
                    uuid={row.id}
                    value={row.comments}
                    onChange={formatAction(actions.updatePrescriptionComments)}
                    placeholder='e.g. take with food'
                    aria-label='Prescription-Comment'
                    className='expanded-input'
                />
            </div>
        </>
    );

    return (
        <BaseCategoryForm<PlanPrescriptionFlat>
            mobile={mobile}
            category='prescriptions'
            categoryData={categoryData}
            numColumns={3}
            addRowLabel='prescription'
            addRow={formatAction(actions.addPrescription)}
            components={{
                gridColumn,
                gridHeaders,
                mobileTitle,
                mobileContent,
            }}
        />
    );
};

export default connect(
    (state: CurrentNoteState, ownProps: CategoryFormOwnProps) => ({
        categoryData: selectPlanCondition(state, ownProps.conditionId)
            .prescriptions,
    }),
    {
        addPrescription,
        updatePrescriptionComments,
        updatePrescriptionDose,
        updatePrescriptionType,
        updatePrescriptionSignature,
    }
)(PrescriptionsForm);
