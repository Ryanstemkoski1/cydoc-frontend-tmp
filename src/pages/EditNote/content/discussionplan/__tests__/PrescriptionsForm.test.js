import Adapter from '@cfaester/enzyme-adapter-react-18';
import Enzyme, { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { PLAN_ACTION as TYPES } from '@redux/actions/actionTypes';
import PrescriptionsForm from '../forms/PrescriptionsForm';
import { categoryId, conditionId, initialPlan } from '../util';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore([]);

const mountWithStore = (
    initStore = { discussionPlan: initialPlan },
    { ...props } = {}
) => {
    const store = mockStore(initStore);
    props = {
        conditionId,
        // Need to mock actual implementation as this function is responsible
        // for action dispatching
        formatAction: jest.fn(
            (action) =>
                (_, { uuid, value }) =>
                    action(conditionId, uuid, value)
        ),
        ...props,
    };
    return {
        store,
        wrapper: mount(
            <Provider store={store}>
                <PrescriptionsForm {...props} />
            </Provider>
        ),
    };
};

describe('PrescriptionsForm', () => {
    it('renders without crashing', () => {
        const { wrapper } = mountWithStore();
        expect(wrapper).toBeTruthy();
    });

    it('matches snapshot', () => {
        const { wrapper } = mountWithStore();
        expect(wrapper.html()).toMatchSnapshot();
    });

    // // TODO: Fix below tests
    // it('dispatches correct action when selecting type from dropdown', () => {
    //     const { store, wrapper } = mountWithStore();
    //     // Open dropdown and select last dropdown option
    //     wrapper
    //         .find('input[aria-label="Prescription-Dropdown"]')
    //         .first()
    //         .simulate('focus');
    //     wrapper
    //         .find('.dropdown__control--is-focused')
    //         .first()
    //         .simulate('mousedown');
    //     const option = wrapper.find('.option').first();
    //     option.simulate('click');

    //     const expectedActions = [
    //         {
    //             type: TYPES.UPDATE_PRESCRIPTION_TYPE,
    //             payload: {
    //                 conditionIndex: conditionId,
    //                 prescriptionIndex: categoryId,
    //                 newType: option.prop('value'),
    //             },
    //         },
    //     ];
    //     expect(store.getActions()).toEqual(expectedActions);
    // });

    // it('dispatches correct action when adding type not found in dropdown', () => {
    //     const { store, wrapper } = mountWithStore();

    //     const value = 'MOST CERTAINLY NOT A MEDICATION';
    //     const input = wrapper.find('input[aria-label="Prescription-Dropdown"]');
    //     input.instance().value = value;
    //     input.simulate('change', { target: { value } });
    //     input.simulate('keyDown', { keyCode: 9, key: 'Tab' }); // validates change

    //     const expectedActions = [
    //         {
    //             type: TYPES.UPDATE_PRESCRIPTION_TYPE,
    //             payload: {
    //                 conditionIndex: conditionId,
    //                 prescriptionIndex: categoryId,
    //                 newType: value,
    //             },
    //         },
    //     ];
    //     expect(store.getActions()).toEqual(expectedActions);
    // });

    // it('dispatches correct action when updating dose', () => {
    //     const { store, wrapper } = mountWithStore();

    //     const value = 'foo';
    //     wrapper
    //         .find('input[aria-label="Prescription-Amount"]')
    //         .simulate('change', {
    //             target: { value },
    //         });
    //     const expectedActions = [
    //         {
    //             type: TYPES.UPDATE_PRESCRIPTION_DOSE,
    //             payload: {
    //                 conditionIndex: conditionId,
    //                 prescriptionIndex: categoryId,
    //                 newDose: value,
    //             },
    //         },
    //     ];
    //     expect(store.getActions()).toEqual(expectedActions);
    // });

    it('dispatches correct action when updating signature', () => {
        const { store, wrapper } = mountWithStore();

        const value = 'foo';
        wrapper
            .find('[aria-label="Prescription-Signature"]')
            .first()
            .simulate('change', {
                target: { value },
            });
        const expectedActions = [
            {
                type: TYPES.UPDATE_PRESCRIPTION_SIGNATURE,
                payload: {
                    conditionIndex: conditionId,
                    prescriptionIndex: categoryId,
                    newSignature: value,
                },
            },
        ];
        expect(store.getActions()).toEqual(expectedActions);
    });

    it('dispatches correct action when updating comment', () => {
        const { store, wrapper } = mountWithStore();

        const value = 'foo';
        wrapper
            .find('[aria-label="Prescription-Comment"]')
            .first()
            .simulate('change', {
                target: { value },
            });
        const expectedActions = [
            {
                type: TYPES.UPDATE_PRESCRIPTION_COMMENTS,
                payload: {
                    conditionIndex: conditionId,
                    prescriptionIndex: categoryId,
                    newComments: value,
                },
            },
        ];
        expect(store.getActions()).toEqual(expectedActions);
    });

    // it('dispatches correct action when adding row', () => {
    //     const { store, wrapper } = mountWithStore();

    //     wrapper.find('button[aria-label="add-row"]').simulate('click');
    //     const expectedActions = [
    //         {
    //             type: TYPES.ADD_PRESCRIPTION,
    //             payload: {
    //                 conditionIndex: conditionId,
    //             },
    //         },
    //     ];
    //     expect(store.getActions()).toEqual(expectedActions);
    // });
});
