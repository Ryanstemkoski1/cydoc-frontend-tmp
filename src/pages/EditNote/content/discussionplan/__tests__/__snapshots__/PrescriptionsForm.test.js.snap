// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrescriptionsForm matches snapshot 1`] = `"<div class=\\"accordion ui fluid plan-section\\"><div class=\\"active title\\"><h2 class=\\"ui large attached header\\">Prescriptions</h2></div><div class=\\"content active\\"><div class=\\"ui stackable three column grid section-body\\"><div class=\\"row\\"><div class=\\"column\\"> Rx </div><div class=\\"column\\"> Signature (Sig) </div><div class=\\"column\\"> Comments </div></div><div class=\\"column\\"><div class=\\"container\\" id=\\"prescriptions-main-input-container\\"><div class=\\"main-input recipe css-2b097c-container\\"><span aria-live=\\"polite\\" aria-atomic=\\"false\\" aria-relevant=\\"additions text\\" class=\\"css-1f43avz-a11yText-A11yText\\"></span><div class=\\"dropdown__control css-yk16xz-control\\"><div class=\\"dropdown__value-container css-g1d714-ValueContainer\\"><div class=\\"dropdown__placeholder css-1wa3eu0-placeholder\\">medication name</div><div class=\\"css-b8ldur-Input\\"><div class=\\"dropdown__input\\" style=\\"display: inline-block;\\"><input autocapitalize=\\"none\\" autocomplete=\\"off\\" autocorrect=\\"off\\" id=\\"react-select-3-input\\" spellcheck=\\"false\\" tabindex=\\"0\\" type=\\"text\\" aria-autocomplete=\\"list\\" aria-label=\\"Prescription-Dropdown\\" style=\\"box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;\\" value=\\"\\"><div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\"></div></div></div></div><div class=\\"dropdown__indicators css-1hb7zxy-IndicatorsContainer\\"><span class=\\"dropdown__indicator-separator css-1okebmr-indicatorSeparator\\"></span><div class=\\"dropdown__indicator dropdown__dropdown-indicator css-tlfecz-indicatorContainer\\" aria-hidden=\\"true\\"><svg height=\\"20\\" width=\\"20\\" viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" class=\\"css-tj5bde-Svg\\"><path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\"></path></svg></div></div></div></div><div class=\\"container\\" id=\\"prescriptions-container3\\"></div><label>Prescription Dose</label><div class=\\"container\\" id=\\"prescriptions-container2\\"></div><div class=\\"ui form\\"><textarea type=\\"text\\" uuid=\\"category\\" aria-label=\\"Prescription-Amount\\" placeholder=\\"e.g. 81 mg tablet\\" class=\\"recipe-amount lg\\" rows=\\"3\\"></textarea></div></div></div><div class=\\"column\\"><div class=\\"ui form\\"><textarea uuid=\\"category\\" aria-label=\\"Prescription-Signature\\" placeholder=\\"e.g. 1 tablet every 8 hours\\" rows=\\"3\\"></textarea></div></div><div class=\\"column\\"><div class=\\"ui form\\"><textarea uuid=\\"category\\" aria-label=\\"Prescription-Comment\\" placeholder=\\"e.g. take with food\\" rows=\\"3\\"></textarea></div></div></div><div class=\\"add-row\\"><button aria-label=\\"add-row\\" class=\\"ui circular icon button hpi-ph-button\\"><i aria-hidden=\\"true\\" class=\\"plus icon\\"></i></button>add prescription</div></div></div>"`;
